UCT: Introduce uct_md_attr_t and uct_md_mem_query

Akshay-Venkatesh opened the pull request:
## What
Add uct_md_mem_query function. Revised https://github.com/openucx/ucx/pull/4943 based on @yosefe feedback

## Why
Today uct_memory_detect_md returns just the memory type associated with the pointer passed to it. We need a more generic return type to capture both memory type and detect the system device type associated with buffer. This allows us to make topology-aware decisions in protocols to specific to memory type and the device backing the buffer.

swx-jenkins3 commented:
Can one of the admins verify this patch?

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,55 @@ struct uct_md_attr {
+ * are present.
+ */
+enum uct_md_mem_attr_field {
+    UCT_MD_MEM_ATTR_MEM_TYPE = UCS_BIT(0), /**< Indicate if memory type is populated. Types
```
UCT_MD_MEM_ATTR_FIELD_MEM_TYPE

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,55 @@ struct uct_md_attr {
+enum uct_md_mem_attr_field {
+    UCT_MD_MEM_ATTR_MEM_TYPE = UCS_BIT(0), /**< Indicate if memory type is populated. Types
+                                                belong to one of @ref ucs_memory_type_t */
+    UCT_MD_MEM_ATTR_SYS_DEV  = UCS_BIT(1)  /**< Indicate if details of system device backing
```
UCT_MD_MEM_ATTR_FIELD_SYS_DEV

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,55 @@ struct uct_md_attr {
+ * pointer depending on the bit fields populated in field_mask
+ */
+typedef struct uct_md_mem_attr {
+    uint64_t          field_mask; /**< @ref uct_md_mem_attr_t */
```
pls expand a bit more on the usage of field_mask ,see https://github.com/openucx/ucx/blob/master/src/uct/api/uct.h#L1142 for example

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+     */
+    uint64_t          field_mask;
+    ucs_memory_type_t mem_type;   /**< Is the type CPU memory or GPU memory, etc*/
+    ucs_sys_device_t  sys_dev;    /**< location of device pointer. eg: NUMA/GPU */
```
location -> Location

Akshay-Venkatesh replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+ *
+ * This structure defines the attributes of a memory pointer which may
+ * include memory type of the pointer, the system device that backs the
+ * pointer depending on the bit fields populated in field_mask
```
end with period

Akshay-Venkatesh replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+     * @ref uct_md_mem_attr_t. Note that the field maske is
+     * populated upon return from uct_md_mem_query and not set by user.
+     * Subsequent use of members of the structure are valid after ensuring that
+     * relevant bits in the field_mask rare set
```
end with period

Akshay-Venkatesh replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+typedef struct uct_md_mem_attr {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_md_mem_attr_t. Note that the field maske is
```
maske -> mask

Akshay-Venkatesh replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+     * relevant bits in the field_mask rare set
+     */
+    uint64_t          field_mask;
+    ucs_memory_type_t mem_type;   /**< Is the type CPU memory or GPU memory, etc*/
```
pls add space line between fields (see other structs in this file for example layout)

Akshay-Venkatesh replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+ * @brief Query attributes of a given pointer
+ *
+ * Return attributes such as memory type, and system device for the
+ * given pointer of specific length. @ref uct_md_mem_attr_t
```
remove the ref, it would be linked anyway because of param type

Akshay-Venkatesh replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+ * given pointer of specific length. @ref uct_md_mem_attr_t
+ *
+ * @param [in]     md          Memory domain to allocate memory on.
+ * @param [in]     address     The address of the pointer. Must be non-NULL
```
end with period

Akshay-Venkatesh replied:
done

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1214,66 @@ struct uct_md_attr {
+ * @param [in]     md          Memory domain to allocate memory on.
+ * @param [in]     address     The address of the pointer. Must be non-NULL
+ * @param [in]     length      Indicates the range of memory starting from (address) to
+ *                             (address + length) bytes. Must be >0.
```
Length of memory region to examine. Must be nonzero.

Akshay-Venkatesh replied:
done

yosefe commented:
@Akshay-Venkatesh
1. see doxygen error in AZP: `/scrap/azure/agent-01/AZP_WORKSPACE/1/s/src/uct/api/uct.h:1243: error: unable to resolve reference to `ucs_memory_type_t' for \ref command (warning treated as error, aborting now)`'

2. pls squash all since this is small PR

3. pls add "Why" to the PR desc

4. @shamisp @tonycurtis pls review as well

yosefe commented on src/ucs/memory/memory_type.h:
```diff
@@ -21,6 +21,7 @@ BEGIN_C_DECLS
 
 
 /*
+ * @ingroup UCS_RESOURCE
```
you need to add proper doxygen doc for ucs_memory_type_t

mellanox-github commented:
<details><summary>Mellanox CI: <b>FAILED</b> on 5 of 17 workers (click for details)</summary>

**Note:** the logs will be deleted after 06-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/885b8e81763973a980d5b8caafe25dd0405dbfcf/_main) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/327c378af154b615e488262bfa0b8ac0c9dfe14d/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/24978e7314b3080275dbb9edfaa2f465123b19d8/r-vmb-ppc-jenkins_W1) | :x: FAILURE
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/aea6d3c048f1f8aef9d459904c1b1892ba2029fc/r-vmb-ppc-jenkins_W2) | :x: FAILURE
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/5c066110b5af68f23cc2063bb01be3e32aeba117/r-vmb-ppc-jenkins_W3) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/6833cd80d89c9e5d2b16ef4a0eac9d86b798fc60/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/cc9b838b1c0c4a7be658e4b03faa3126a8f8f558/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/5304086f3ee06cdb84ab864f3228675db5f8c1e0/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/b75ef277f0ef2de571636dfbe3864e1f05b04ed9/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/b0d883ffa454e8b00a62f3c248c09c333b8124fb/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/a5a26f967b1c71475ffb3cc1ae8f5d8a00e817aa/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/a56e174f3ef84aae76a897b557f2a046f865ce90/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/0e4dac488093f5862c3e66cc87572eb13dab38b5/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/1741e7c00fcd531f918e6e10eed6fc9cadee430b/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/2298d5111ee9e1497a9cf37b0edb7d0535cde82b/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/3ce0e2603a142a68c0d3b17da51dc809afb4439e/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/c086baeb3fbcb7d950b0d38e099dcd70/raw/74df391b028c041ab2e45b8af4ae725027884d3d/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS

mellanox-github commented:
<details><summary>Mellanox CI: <b>FAILED</b> on 10 of 17 workers (click for details)</summary>

**Note:** the logs will be deleted after 06-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/451daa90309577f180ae346409d2d534d5183776/_main) | :x: FAILURE
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/536414aa929fc83937645ba3fee5fd3ed99f375c/hpc-arm-hwi-jenkins_W0) | :x: FAILURE
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/15b931cfe85f123fe8d420167b75ad769f631ac9/hpc-arm-hwi-jenkins_W1) | :x: FAILURE
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/1623660166cfdfc29a392a8d854d5439a4fa0f12/hpc-arm-hwi-jenkins_W2) | :x: FAILURE
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/0645c64d84a8e5987650e539f508c284b78df5fa/hpc-arm-hwi-jenkins_W3) | :x: FAILURE
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/cd653e65410f92fd88658d1e764e8882bb705958/hpc-test-node-new_W0) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/120d547e61feb7961abe2cbde582cb0d2a9c7487/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/ffed38a48782138098ec80adb67cc3813b508920/r-vmb-ppc-jenkins_W1) | :x: FAILURE
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/917cc52f71728be5206ce319e1c7182ac9817d6f/r-vmb-ppc-jenkins_W2) | :x: FAILURE
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/2d4b1fdbd54c1e4efa7b35440b37c660330519c7/r-vmb-ppc-jenkins_W3) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/967c2ede4b5bbf3b6937f3b63a8cfb3008363d3a/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/18ab806fce49503cb09616bd898dbf086fab7a0a/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/1ffb92ef77fffc302af9a556b22e3c328ce6ba5b/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/0ae6687bef2ce88101638bb5eb5c683a4b812d8f/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/c260afe07ae2dd8be5ace180fb71510fe63494d7/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/9ff76561fbe9d11160d847c773e3131197ef5a35/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/186b3a8a1c259fb3e135fbbc36af1e8c/raw/6c98dbe290be8e3ed249079f27c8ad56331396ce/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS

mellanox-github commented:
<details><summary>Mellanox CI: <b>PASSED</b#  > on 17 workers (click for details)</summary>

**Note:** the logs will be deleted after 06-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/8ec06ebdb88a2476fcfd57f6a1eeded1eec34dc1/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/3ea84bed8da9dd2b8c76c6dd977e54057f914204/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/29f2a51ba649fcf90d8b1225226d6e84f1f5ac39/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/72760d5d51d78faf6477f3fe46613ec682bdf7a4/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/683ba38be4f073c4cf8018120e827176c323c48b/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/b0ebd969969a83bd4e3e24a27b7de03ac9cb07f0/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/e4b142e3dafc3d052ce490dbf27f08b7fe0111c2/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/19b80fb3f0074b5d2b03d301d51bf91e613a6719/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/f10ae2b29a4c1871306b3b67ebc78e2b5e647a63/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/8e6613ee3cdd1f8d7c4d59ee323f91278d34a8b5/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/59723473c4f1499853c1ccf207e4a46090d20306/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/2c74bd7b90836be914305bc955f0fa8926876650/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/b42a157bf6341176547eff233b0cae8ca41faf88/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/bd4dd45c59479b368fe226f9e241a1bbfd491e34/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/684f33b467b4c67d3bfd6702bd7cb62fa8e69b2b/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/f631abc4563fd98f4a344362297814d15abaf942/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/2ba3603f8f79b6e7dd95e7322c3f3612/raw/07d6074aa94098daf2035d27d7337fa47a86e53f/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

mellanox-github commented:
<details><summary>Mellanox CI: <b>PASSED</b#  > on 17 workers (click for details)</summary>

**Note:** the logs will be deleted after 06-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/83273856ec2f2a11da06f414cc56dcf8aab2462a/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/ff092c10f76dfc144574b1e1f01104b23a226946/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/5679483c3953ee6f9d6ccb42d55c2ba37cec732b/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/158a013d4c70fd862d927efeb94f6805798a08d2/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/ddaf794e5dedc28d2f43a7e1e89ed860a38dc090/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/c456a8d52a0b15eaab1177ef60b024759a529ce5/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/f1927e4ecfc29bbd4d65be4fe4d74be4d24dad35/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/5bf4fb42235cb0abbb41fbd62fb56bcc38229483/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/fea3b6333ccd1e3e29d03f1744fc887edc7ad83d/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/37fc90b2e98b923896cd973e0d4f667fc0f0b5d4/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/2eb3690e0575a58f07d72cdbe12f54805bb48ad3/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/32db80bd75d943ff603b55e8c674ef3f03c9daad/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/f34710ccdee4653e7db14d022464f89f3b423730/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/4e9c4f8d57c012e4fe9ffa2f3129b21bbb9c6209/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/e73fc63aae4e9b62cf1744783e4feeda09a920d1/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/25425c9004993270f3a298a9c820e403fe028cf6/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/b5463f8b6bb16024c6183470d41aef7c/raw/ddbe63bf3b4e7397a2bdd3edbd5039631507ae93/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

mellanox-github commented:
<details><summary>Mellanox CI: <b>PASSED</b#  > on 17 workers (click for details)</summary>

**Note:** the logs will be deleted after 06-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/982db898f25218f8d0087d552326a01831af63fe/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/26bb07971beca0da673af1107ee8779a5c368dd9/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/0e9a40fa5b43e27a1dcc3b155d498d92a78954b5/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/88a29f6738d0d1aed1de9642902ec3346276d6af/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/09427ca6a00d42379b7b6021a0c86347ecfdd9df/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/515cd070d8ea94c319ea28000b472abe19573f24/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/d2eb9ee724e4983990f0425f7811bef14219a309/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/843ccda8b5367b3e0ca4af85d6e911ad34d2bed5/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/9f409c6a5b27e633599a2f1dd6ae5e87ec699cca/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/42c28d7cc4a51effb5abc2fb6711de44000f9fcc/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/fb3256a7981f4015176be51e035b77f1d1ceb2a0/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/9938d025a1d697b2e22c8a31467933ee199c9aec/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/807e6f4ba6814752a3c1af11e2506dd88a898fdd/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/cfa40f1a05341b64115d1dc7f9b06b965afc2cd7/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/967efb29b7c3834e0d2706cbb370310b0dfdc3b1/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/a25f3250db5d0509e0f9ad4ad71780b78b197e58/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/55ecd2e60f675fd8fa663afdfbcec717/raw/73ecd489eae379f559175b902751c31ee01e3676/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

brminich commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1215,72 @@ struct uct_md_attr {
+ * Return attributes such as memory type, and system device for the
+ * given pointer of specific length.
+ *
+ * @param [in]     md          Memory domain to allocate memory on.
```
I'd rephrase somehow to reflect that this function does not allocate memory. E.g, like below:
`Memory domain the memory was allocated on`?

Akshay-Venkatesh replied:
Thanks for the catch @brminich

This was a copy paste error. I'll rephrase this

Akshay-Venkatesh replied:
@brminich 
I want to avoid `MD where memory was allocated on` because if cuda pointers are queried with this function, then it's not necessary that UCT cuda md was used to allocate. I think the same argument applies to host pointers. The query function would return failure for pointer queries on MDs that don't recognize the pointer kind.

Does the following work instead?
```
 * @param [in]     md          Memory domain to run query on. 
                                           This function returns error if MD does not recognize the pointer.
```

brminich replied:
👍

brminich replied:
@Akshay-Venkatesh, looks good to me. Thanks

mellanox-github commented:
<details><summary>Mellanox CI: <b>PASSED</b#  > on 13 workers (click for details)</summary>

**Note:** the logs will be deleted after 07-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/e7d5fe26ee170e81384640a597249b0659ac5152/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/d7437aba244967d9870808cf7219ed4cf9c6199f/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/dff74f4bec50285ad4755da0ac00f76d711724be/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/52644f0251bcafd35043d3ae9e19d03400bd769c/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/6390d055429f27ec2082c7c165bb7822a30872f4/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/7c37266e8eb3abae158bc384cfc8653eb6990664/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/c99636777c718710aff7e2162fd47168ca08d8a7/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/3537406fefb5efc226e8a44d6bfc016647daa0ca/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/61a0fa38782289bbb4547cb5c6fc032c711c5cdd/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/7d232fcc3bb73aefbb517085a70c94d98f31bbe8/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/a658e91aea1221162366bb1fbdb80d4912928f62/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/6ae1be09fdbedbf69ef17709df8efa7e78377a27/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/18f75a4f2169534d6db14f1463150c34/raw/6195dd5a3c8ba334c17d1e02fc6e1bae5a6479ae/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

Akshay-Venkatesh commented:
ping @shamisp @tonycurtis 

Just in case you missed the review request

Akshay-Venkatesh commented:
> 
> 
> uct.h:1227 and others: Eg -> E.g.
> 1228-1321: details ... are populated
> 1240: the memory type ... and the system device....
> 1254: Drop the "is". Is it clear to all readers what "etc" enumerates? ("etc.")

Thanks @tonycurtis I've incorporated the changes and have dropped "etc" and added that there would some valid memory type returned instead. Let me know if the changes look good.

mellanox-github commented:
<details><summary>Mellanox CI: <b>UNKNOWN</b> on 5 workers (click for details)</summary>

**Note:** the logs will be deleted after 07-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/cb23a12094b93d3b3a9ee2280ad148dc/raw/4487ec24131b7a618ac92065251001c7d6379f71/_main) | :question: ABORTED
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/cb23a12094b93d3b3a9ee2280ad148dc/raw/6efed103d4936584b3b40b41ffa33e990a7de3d5/hpc-arm-cavium-jenkins_W0) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/cb23a12094b93d3b3a9ee2280ad148dc/raw/b53354f3314b7ba3c65d9165eeee580bd1f0de06/hpc-arm-cavium-jenkins_W1) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/cb23a12094b93d3b3a9ee2280ad148dc/raw/63cc0e2b0ef8f2d4cb60d8418ce0f68183d434b4/hpc-arm-cavium-jenkins_W2) | :question: UNKNOWN
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/cb23a12094b93d3b3a9ee2280ad148dc/raw/cc9397b0cb6e94affbc4a95cc40b586e1228c9e8/hpc-arm-cavium-jenkins_W3) | :question: UNKNOWN

mellanox-github commented:
<details><summary>Mellanox CI: <b>PASSED</b#  > on 9 workers (click for details)</summary>

**Note:** the logs will be deleted after 07-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/6bf4d004c79baf815c6c70c0b7b2af9bae371919/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/a0e36bac0a7cf2914a359889cd257f1171cc9d06/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/063d3328043333ca4fea6ddb8bb84583a8c9e644/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/2ded860fd4bd028d7dd30c2ceab77b9cfb16faaa/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/18d0efc8b7090cffeba396d195a89068169f6df9/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/4645e24c1cd451359f4f40eb2a9b2850c6f2a05c/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/26e8044ba1e18f174a3d3c1fc9ce794233b0ed2f/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/a7421e9b1d22fbf14cc968aee304fe131d560104/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/655f378a47b0709f69df94f147ac5041/raw/a981a84d4618c4a19068e2f85385e07e9cbde8b5/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

shamisp commented on src/uct/api/uct.h:
```diff
@@ -1213,6 +1215,74 @@ struct uct_md_attr {
+ * @param [in]     address     The address of the pointer. Must be non-NULL.
+ * @param [in]     length      Length of the memory region to examine.
+ *                             Must be nonzero.
+ * @param [out]    mem_attr    If successful, filled with ptr attributes.
```
what happens if zero is passed to the address or len ?

Akshay-Venkatesh replied:
@shamisp If either is true then `md_mem_query` should return an error. Is it better if that was explicitly stated? I can push that change if so.

shamisp replied:
I would suggest to be more specific and define what error returned in this case. IMHO there are many code that don't have null check before calling function like this. Maybe instead of error we shell return status that the memory and device are undetected ?

Akshay-Venkatesh replied:
@shamisp Looking at `enum ucs_status_t` would returning UCS_ERR_UNSUPPORTED work or is this error reserved something else? If returning this error works, then I'll reference this for NULL/zero arguments.

shamisp replied:
UCS_ERR_INVALID_PARAM sounds like more appropriate. Please documents this explicitly. thanks

Akshay-Venkatesh replied:
Does this look good?
```diff
diff --git a/src/uct/api/uct.h b/src/uct/api/uct.h
index bf59aadf1..816e39737 100644
--- a/src/uct/api/uct.h
+++ b/src/uct/api/uct.h
@@ -1272,9 +1272,11 @@ typedef struct uct_md_mem_attr {
  * @param [in]     md          Memory domain to run the query on. This function
  *                             returns an error if the md does not recognize the
  *                             pointer.
- * @param [in]     address     The address of the pointer. Must be non-NULL.
+ * @param [in]     address     The address of the pointer. Must be non-NULL
+ *                             else UCS_ERR_INVALID_PARAM error is returned.
  * @param [in]     length      Length of the memory region to examine.
- *                             Must be nonzero.
+ *                             Must be nonzero else UCS_ERR_INVALID_PARAM error
+ *                             is returned.
  * @param [out]    mem_attr    If successful, filled with ptr attributes.
  *
  * @return Error code.
```

Akshay-Venkatesh replied:
@shamisp If nothing else is pending, would you mind approving?

Akshay-Venkatesh commented:
@shamisp @tonycurtis  If nothing else is pending, would you mind approving?
